torch/_inductor/codegen/cpp.py

```diff
@@ -3782,48 +3824,141 @@ def can_fuse_vertical(self, node1, node2):
+        generated_cpp_vec_kernel_count = metrics.generated_cpp_vec_kernel_count
+        cpp_kernel_proxy_list: List[CppKernelProxy] = []
+        nodes_list: List[List[SchedulerNode]] = []
+        assert isinstance(node, OuterLoopFusedSchedulerNode)
+
+        def try_outer_loop_fusion_with_local_buf(node: OuterLoopFusedSchedulerNode):
+            """
+            Codegen code with fused outer loop and local Buffer.
+            """
+            assert isinstance(node, OuterLoopFusedSchedulerNode)
+            cpp_kernel_proxy_list.clear()
+            nodes_list.clear()
+
+            def get_call_ranges(node: BaseSchedulerNode):
+                assert isinstance(node, (SchedulerNode, FusedSchedulerNode))
+                nodes: List[SchedulerNode] = node.get_nodes()  # type: ignore[assignment]
+                _, (group, reduction_group) = max(
+                    nodes, key=lambda x: int(x.is_reduction())
+                ).group
+                call_ranges = tuple(group) + tuple(reduction_group)
+                return call_ranges
+
+            LocalBuffer = namedtuple("LocalBuffer", ["local_buf", "global_buf"])
+            local_buffers: List[LocalBuffer] = []
+            if all(
+                len(get_call_ranges(_node)) == node.outer_loop_fusion_depth + 1
+                for _node in node.get_outer_nodes()
             ):
-                # Merge the cpp_kernel_proxy_list into cpp_kernel_proxy
+                # Ref to the typical case of local buffer
+                # in https://github.com/pytorch/pytorch/blob/
+                # 1115a25c36340554442f28f9570abd42f0aface2/aten/src/ATen/native/cpu/SoftMaxKernel.cpp#L159
+                # where the buffer is with size of last dim and contiguous.
+                # Only support this typical case at first.
+                for scheduler_node in node.get_nodes():
+                    # all users inside same OuterLoopFusedSchedulerNode
+                    if not scheduler_node.is_reduction() and all(
+                        user.node in node.get_nodes() for user in scheduler_node.users
+                    ):
+                        global_buffer = scheduler_node.node
+                        assert isinstance(global_buffer, ir.ComputedBuffer)
+                        global_buffer_layout = global_buffer.get_layout()
+                        if not global_buffer_layout.is_contiguous():
+                            continue
+                        # Local Buffer is a view of global buffer
+                        size_offset = node.outer_loop_fusion_depth - len(
+                            get_call_ranges(scheduler_node)
+                        )
+                        local_buffer_layout = ir.FixedLayout(
+                            global_buffer_layout.device,
+                            global_buffer_layout.dtype,
+                            global_buffer_layout.size[size_offset:],
+                            global_buffer_layout.stride[size_offset:],
```

Comment:
I see, so the buffer size is reduced here. Do we know for sure though that the access to the buffer are going to be a simple contiguous order and that the loop dimensions match the number of dimensions in the layout?

Reply:
Thanks for the comment. Previously, I only checked if the `global buffer` is contiguous. I think we do need to verify the buffer access order. I've added some new checks:

https://github.com/pytorch/pytorch/blob/71284cdc72e36bf6ed97c3b6ea3fb5c483e92a6a/torch/_inductor/codegen/cpp.py#L3882-L3895

- For the dimensions in the fused outer loop, where loops may be collapsed, I check the `sympy_product` result between the `global buffer layout` and `schedulerNode's group[1][0]`.

- For the dimensions inside, since we replace these dimensions with the local buffer, there should be no loop collapse and the number of dimensions for loop should match to number of dimensions for layout. Therefore, I check that each size value is the same between the `global buffer layout` and `schedulerNode's group[1][0]`.

Do you think these checks are sufficient to ensure that the buffer access is in contiguous order and matches the number of loop dimensions? Are there any other checks we should add here?

Reply:
Thanks for the comment, adding an util function of `is_all_write_read_contiguous` to check all the `read_index_expr` and `write_index_expr` to this buffer is contiguous. Could you kindly help to take a look again?

One thing to note is that, in the implementation, I have assumed that for the `index_var` in `index_expr`, the one with larger suffix means at the inner dimension. For example, in an `index_expr` of `1024*z0 + 512*z1 + z2`, assumed `z0` is the `outermost` dimension and `z2` is the `innermost` dimension. Do you think this assumption is sound? If not, could you give some more suggestions for how to check the index expr is contiguous?